### PR TITLE
[Perf] add SeaCache algorithm for diffusion acceleration

### PR DESCRIPTION
**Pull request by zhooooong**
# Purpose

Adds SeaCache ([CVPR 2026](https://arxiv.org/abs/2602.18993)) as a new diffusion cache backend (`cache_backend="sea_cache"`), wired into three DiT families: FLUX.1-dev, FLUX.2-klein, and the Qwen-Image family (`QwenImageTransformer2DModel`: Qwen-Image, Qwen-Image-Edit, Qwen-Image-Edit-2511).

SeaCache keeps TeaCache's accumulate-and-refresh rule but measures the step-to-step distance *after* a timestep-dependent Wiener filter, so the metric tracks content change instead of noise. It needs no per-checkpoint fitted coefficients — one threshold controls the whole trade-off.

```bash
vllm-omni serve black-forest-labs/FLUX.1-dev --omni --port 8091 --cache-backend sea_cache --cache-config '{"sea_thresh": 0.3}'
```

On FLUX.1-dev at 1024x1024 / 50 steps, `sea_thresh=0.3` cuts generation latency **57.6%** (6.98 s -> 2.96 s) at 30.19 dB PSNR against the uncached output. On Qwen-Image-Edit-2511 at 1024x1024 / 40 steps it cuts **53.4%** (23.48 s ->
10.94 s) at 40.04 dB.

## How it works

Diffusion establishes low-frequency structure early and refines high-frequency detail later, so the band the denoiser is working on moves as sampling proceeds. Per step, SeaCache builds the Wiener response of the optimal linear denoiser at the current noise level, applies it to the timestep-modulated input of transformer block 0, and accumulates the relative L1 distance between consecutive filtered features. Below the threshold the block stack is skipped and the cached residual is added back; at or above it the blocks run and the accumulator resets. First and last step always run.

Implementation follows the existing TeaCache hook architecture — **zero changes to model code**. The new `vllm_omni/diffusion/cache/seacache/` subpackage holds the SEA filter math (ported to match the reference exactly) and a hook that mirrors `TeaCacheHook` (CFG-branch-aware state, residual reuse); all model-specific logic lives in the extractor registry via optional new `CacheContext` fields.

## Performance

FLUX.1-dev, 1024x1024, 50 steps, 15 DrawBench-style prompts, fixed per-prompt seeds, 1x H100 (80GB), offline `Omni` engine, one engine per config, 1 warmup generation excluded.

| Config             | Latency | vs baseline | PSNR  | SSIM  | LPIPS |
| --------------------| --------:| ------------:| ------:| ------:| ------:|
| baseline           | 6.98 s  | -           | -     | -     | -     |
| `sea_thresh=0.215` | 3.80 s  | -45.5%      | 35.74 | 0.968 | 0.025 |
| `sea_thresh=0.30`  | 2.96 s  | **-57.6%**  | 30.19 | 0.934 | 0.063 |
| `sea_thresh=0.42`  | 2.55 s  | -63.5%      | 29.24 | 0.921 | 0.082 |
| `sea_thresh=0.60`  | 2.00 s  | -71.3%      | 24.13 | 0.861 | 0.168 |

VRAM is unchanged by the cache: SeaCache adds one cached residual plus transient FFT buffers (~10-30 MB).

### vs TeaCache: SeaCache wins decisively

TeaCache swept at three thresholds, each matched by a SeaCache threshold tuned onto the same latency:

| Latency (tea / Sea)           | PSNR: tea / Sea | dPSNR (paired)    | SeaCache better on |
| -------------------------------| -----------------| -------------------| --------------------|
| 3.85 s / 3.80 s (`sea 0.215`) | 24.61 / 35.74   | **+11.13 ± 1.19** | 15/15 prompts      |
| 2.65 s / 2.68 s (`sea 0.38`)  | 20.97 / 29.67   | **+8.70 ± 1.07**  | 14/15 prompts      |
| 2.17 s / 2.13 s (`sea 0.522`) | 19.17 / 24.26   | **+5.08 ± 0.86**  | 14/15 prompts      |

### Why: the schedules differ in shape, not just in rate

Refresh pattern for one prompt over 50 steps (`R` = blocks ran, `.` = skipped):

```
teacache 0.25   RR.R.R.R.R.R.R.R.RR.R.R.R.R.R.R.R.R.R.R.R.R.R.R.RR   ratio 0.54
teacache 0.60   RR...R...R...R...R..R...R...R...R...R...R..R..R..R   ratio 0.30
seacache 0.215  RRRRRRRRRR.RR.R.R.R.R..R..R..R..R..R..R..R..R.R.RR   ratio 0.54
seacache 0.60   RR.R.R.R..R...R...R......R......R.......R.....R..R   ratio 0.26
```

TeaCache degenerates to a near-fixed stride, spending half its budget on late steps. SeaCache front-loads (10 of its 27 refreshes at delta=0.215 fall in the first quarter) and its gaps widen monotonically. Early steps set image structure, which is why the same refresh budget buys ~5 dB more.

### vs cache-dit

| Latency (cache-dit / SeaCache)                  | PSNR: cdit / Sea | dPSNR (paired)   | SeaCache better on |
| -------------------------------------------------| ------------------| ------------------| --------------------|
| 3.49 s / 3.44 s (`cdit rdt 0.08` / `sea 0.239`) | 30.27 / 31.04    | +0.78 ± 0.32     | 13/15 prompts      |
| 2.96 s / 2.96 s (`cdit rdt 0.12` / `sea 0.30`)  | 27.47 / 30.19    | **+2.71 ± 0.61** | 14/15 prompts      |
| 2.42 s / 2.41 s (`cdit rdt 0.36` / `sea 0.455`) | 23.67 / 29.12    | **+5.46 ± 0.68** | 15/15 prompts      |

### vs mag-cache

 Latency (mag / Sea)                       | PSNR: mag / Sea | dPSNR (paired)   | SeaCache better on |
| -------------------------------------------| -----------------| ------------------| --------------------|
| 2.78 s / 2.81 s (`mag 0.10` / `sea 0.35`) | 25.74 / 29.87   | **+4.13 ± 0.74** | 15/15 prompts      |
| 2.24 s / 2.29 s (`mag 0.24` / `sea 0.49`) | 23.66 / 27.03   | **+3.37 ± 0.62** | 14/15 prompts      |
| 1.97 s / 2.00 s (`mag 0.40` / `sea 0.60`) | 21.73 / 24.13   | **+2.40 ± 0.93** | 12/15 prompts      |

### Qwen-Image-Edit-2511 validation

Correctness on the second architecture (`QwenImageEditPlusPipeline`, true CFG with a negative prompt, 1024x1024, 40 steps, fixed seeds). Three distinct edit tasks, each with its own input image: recolor a bird (blue -> cardinal), colorize a black-and-white portrait, and repaint an oil painting as a photograph:

| Config              | Latency | vs baseline | Refresh ratio | PSNR              | SSIM  | LPIPS |
| ---------------------| --------:| ------------:| --------------:| ------------------:| ------:| ------:|
| baseline (uncached) | 23.43 s | -           | 1.00          | -                 | -     | -     |
| `sea_thresh=0`      | 23.64 s | +0.9%       | 1.00          | **bit-identical** | 1.000 | 0.000 |
| `sea_thresh=0.15`   | 15.55 s | **-33.6%**  | 0.65          | 44.05 ± 3.11      | 0.998 | 0.004 |
| `sea_thresh=0.30`   | 10.95 s | **-53.3%**  | 0.45          | 40.88 ± 3.10      | 0.994 | 0.012 |

### FLUX.2-klein-4B validation

Correctness on the third architecture (`Flux2KleinPipeline`, guidance-distilled, single branch, 1024x1024, 3 prompts, seed 42).

| Config (10 steps)   | Latency | vs baseline | Refresh | PSNR         | SSIM  | LPIPS |
| ---------------------| --------:| ------------:| --------:| -------------:| ------:| ------:|
| baseline (uncached) | 0.76 s  | -           | -       | -            | -     | -     |
| `sea_thresh=0.60`   | 0.64 s  | -14.7%      | 0.80    | 32.25 ± 1.88 | 0.984 | 0.018 |
| `sea_thresh=1.00`   | 0.53 s  | -30.2%      | 0.60    | 23.10 ± 0.71 | 0.929 | 0.078 |
| `sea_thresh=1.50`   | 0.47 s  | -37.7%      | 0.50    | 20.33 ± 1.23 | 0.874 | 0.145 |

## Test Plan

**vLLM Version:** 0.28.0
**vLLM-Omni Commit:** `d5aced92` (main at time of work)

New unit-test file `tests/diffusion/cache/test_seacache.py` — 19 cases

| Scope                                                              | Result                   |
| --------------------------------------------------------------------| --------------------------|
| `tests/diffusion/cache/` (incl. the 19 new cases)                  | **188 passed** (was 167) |
| `tests/diffusion/models/flux/`                                     | 4 passed                 |
| E2E FLUX.1-dev, 1x H100 — benchmark and gate tables above          | PASS                     |
| E2E Qwen-Image-Edit-2511, 1x H100 — 3 edit tasks, table above      | PASS                     |
| E2E FLUX.2-klein-4B, 1x H100 — 10-step skip table above | PASS                     |

## Test Result

All green. Unit and regression suites ran on CPU (Python 3.12, vLLM 0.28.0, PyTorch 2.13.0+cu130, pytest 9.1.1); E2E runs used the offline `Omni` engine on one NVIDIA H100.

### CODE REVIEW

**Comment by BruceLoveDecimal** [comment]
It seems there is some overlaps with PR #6922 👀

**Comment by MaciejBalaNV** [comment]
Hey @zhooooong, as @BruceLoveDecimal mentioned, there is already a PR that adds SeaCache. Can you please test out the Flux and Qwen-Image model support using that implementation? Then we could either extend https://github.com/vllm-project/vllm-omni/pull/6922 with Flux and Qwen-Image support or wait until it's merged and create a separate PR that adds Flux and Qwen-Image support.

**Comment by zhooooong** [comment]
> Hey @zhooooong, as @BruceLoveDecimal mentioned, there is already a PR that adds SeaCache. Can you please test out the Flux and Qwen-Image model support using that implementation? Then we could either extend #6922 with Flux and Qwen-Image support or wait until it's merged and create a separate PR that adds Flux and Qwen-Image support.

Thanks @BruceLoveDecimal @MaciejBalaNV for flagging this — I hadn't noticed that PR.
I took a look at #6922, and while there is indeed significant overlap with ours, I'd argue the two PRs are complementary rather than conflicting:
 - Model coverage: ours targets image DiTs (FLUX.1-dev, FLUX.2-klein, Qwen-Image family), while PR6922 targets Cosmos3 (video).
 - Fidelity to the paper: our implementation follows the paper's decision feature — filtering the timestep-modulated input of the first transformer block (the same signal TeaCache uses) — as well as plain reuse of the cached residual on skipped steps; PR6922 instead filters the raw noisy latents and replaces residual reuse with polynomial extrapolation (residual_order).
   - Engineering trade-offs: PR6922 includes several targeted enhancements for the video scenario that we'd be glad to adopt, e.g.:
     - a max_consecutive_cached cap;
     - cross-rank synchronization of the skip decision, making it safe under parameter sharding / sequence parallel.
  
The file-level conflicts between the two (the seacache/ package, the sea_cache backend registration, config field names) are mechanical and easy to reconcile.

File: vllm_omni/diffusion/cache/seacache/hook.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+    def new_forward(self, module: torch.nn.Module, *args: Any, **kwargs: Any) -> Any:
+        ctx = self.extractor_fn(module, *args, **kwargs)
+
+        if self._sequence_parallel_active():
+            return self._run_uncached(ctx)
+
+        # Same CFG branch selection as TeaCacheHook: CFG-parallel gives each
+        # rank one branch; otherwise branches alternate (positive first).
+        if getattr(module, "do_true_cfg", False):
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/seacache/hook.py:74]: Set Klein's transformer.do_true_cfg before selecting cache branches.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 2]
```diff
@@ -663,12 +663,14 @@ def _refresh_cache_for_requests(
         if num_inference_steps is None and od_config.cache_backend in (
             "tea_cache",
             "step_cache",
+            "sea_cache",
```
Comment (lishunyang12) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py:666]: Resolve SeaCache's actual step count before refreshing its state.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 3]
```diff
@@ -1105,6 +1157,10 @@ def extract_flux_context(
     # ============================================================================
     # PREPROCESSING (Flux1-specific)
     # ============================================================================
+    # The pipeline passes t / 1000 == the scheduler's shifted sigma
+    # (FlowMatchEuler); capture it before the * 1000 rescale below.
+    sigma = float(timestep.to(device=hidden_states.device, dtype=torch.float32).reshape(-1)[0].item())
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:1162]: Compute SeaCache metadata outside shared TeaCache extractors to avoid per-step GPU synchronization.

File: tests/diffusion/cache/test_seacache.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        cleanup_dist_env_and_memory,
+        init_distributed_environment,
+        initialize_model_parallel,
+    )
+
+    from vllm_omni.diffusion.attention.backends.sdpa import SDPABackend
+
+    os.environ.setdefault("MASTER_ADDR", "localhost")
+    os.environ.setdefault("MASTER_PORT", "29503")
+    init_distributed_environment(world_size=1, rank=0, local_rank=0, distributed_init_method="env://")
```
Comment (lishunyang12) [inline tests/diffusion/cache/test_seacache.py:51]: Use backend='gloo' in this CPU fixture.

File: vllm_omni/diffusion/cache/seacache/backend.py [hunk 5]
```diff
@@ -54,9 +71,44 @@ def enable_cosmos3_seacache(
...
+    "FluxPipeline": extract_flux_seacache_context,
+    "Flux2Pipeline": extract_flux2_seacache_context,
+    "Flux2KleinPipeline": extract_flux2_seacache_context,
+    "QwenImagePipeline": extract_qwen_seacache_context,
+    "QwenImageEditPipeline": extract_qwen_seacache_context,
+    "QwenImageEditPlusPipeline": extract_qwen_seacache_context,
+}
+
+
+def enable_image_seacache(pipeline: Any, config: DiffusionCacheConfig) -> SeaCacheRootHook:
```
Comment (MaciejBalaNV) [inline vllm_omni/diffusion/cache/seacache/backend.py:87]: I think this name can be misleading, since we only enable the cache for Flux and Qwen image models, not all existing image models. I'd mention Flux and Qwen explicitly.

File: vllm_omni/diffusion/distributed/cfg_parallel.py [hunk 6]
```diff
@@ -88,6 +90,48 @@ class CFGParallelMixin(metaclass=ABCMeta):
...
+            if sigma_index is None:
+                sigma_index = (getattr(scheduler, "begin_index", None) or 0) + step_index
+            sigmas = getattr(scheduler, "sigmas", None)
+            self._current_sigma = sigmas[sigma_index] if sigmas is not None else None
+            yield
+        finally:
+            self._current_step_index = None
+            self._current_sigma = None
+
+    def _predict_noise_with_cache_context(self, context_name: str, model_kwargs: dict[str, Any] | None):
```
Comment (MaciejBalaNV) [inline vllm_omni/diffusion/distributed/cfg_parallel.py:124]: Can you elaborate why it's needed? We also have a separate Cosmos3 transfer modality fix for SeaCache here: https://github.com/vllm-project/vllm-omni/pull/7939

File: vllm_omni/diffusion/cache/seacache/backend.py (hunk 5 above)
Comment (zhooooong) [reply vllm_omni/diffusion/cache/seacache/backend.py:87]: > I think this name can be misleading, since we only enable the cache for Flux and Qwen image models, not all existing image models. I'd mention Flux and Qwen explicitly.

The generic name is intentional: `enable_image_seacache` is the shared integration point for image diffusion pipelines, while `_IMAGE_EXTRACTORS` is the capability registry that defines the models currently supported. Flux and Qwen Image are the current entries, but future image models with compatible SeaCache adapters should be added through the same entry point rather than introducing model-family-specific enablers.

File: vllm_omni/diffusion/distributed/cfg_parallel.py (hunk 6 above)
Comment (zhooooong) [reply vllm_omni/diffusion/distributed/cfg_parallel.py:124]: > Can you elaborate why it's needed? We also have a separate Cosmos3 transfer modality fix for SeaCache here: #7939

This helper is placed in `CFGParallelMixin` for image pipelines that use its standard two-branch CFG contract. The mixin is the layer that decides whether the current local forward is the conditional or unconditional branch, in both sequential and CFG-parallel execution, so it is also the correct layer to enter the corresponding cond or uncond cache context. This keeps cache branch ownership out of each individual image pipeline.

Cosmos3 is different even though it also inherits `CFGParallelMixin`: its transfer path has model-specific branches such as `cond_no_control`, so the Cosmos3 pipeline owns those names and passes `_cache_context` explicitly. #7939 addresses Cosmos3 Transfer’s target-only indicator and multi-branch lifecycle; it does not replace the standard image CFG context dispatch here.
